python/lsst/daf/persistence/safeFileIo.py

```diff
@@ -48,6 +50,40 @@ def setFileMode(filename):
     os.chmod(filename, (~umask & 0o666))
 
 @contextmanager
+def FileForWriteOnceCompareSame(name):
+    """Context manager to get a file that can be written only once and all other writes will succeed only if
+    they match the inital write.
+
+    The context manager provides a temporary file object. After the user is done, the temporary file becomes
+    the permanant file if the file at name does not already exist. If the file at name does exist the 
+    temporary file is compared to the file at name. If they are the same then this is good and the temp file
+    is silently thrown away. If they are not the same then a runtime error is raised.
+    """
+    outDir, outName = os.path.split(name)
+    safeMakeDir(outDir)
+    with tempfile.NamedTemporaryFile(dir=outDir, prefix=outName, delete=False) as temp:
+        try: 
+            yield temp
+        finally:
+            try:
+                temp.close()
+                os.symlink(temp.name, name)
+                # If the symlink was created then this is the process that created the first instance of the 
+                # file, and we know its contents match. Move the temp file over the symlink.
+                if os.path.islink(name):
+                    os.rename(temp.name, name)
+            except OSError as e:
+                if e.errno != errno.EEXIST:
+                    raise e
+                filesMatch = filecmp.cmp(temp.name, name, shallow=False)
```

Comment:
There is the possibility that this could get more complicated.  For example, there could be randomization of the order of Config items or headers or something.  We might need to pass in a comparison function in some cases.